components/service/glean/src/main/java/mozilla/components/service/glean/utils/FreshBootDetector.kt

```diff
@@ -0,0 +1,102 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.components.service.glean.utils
+
+import android.content.Context
+import android.os.Build
+import android.provider.Settings
+import android.support.annotation.VisibleForTesting
+import java.io.File
+import java.io.IOException
+import java.util.UUID
+
+internal object FreshBootDetector {
+    // cache the value of the first check so we can safely run it multiple times
+    @VisibleForTesting
+    var cacheIsFreshBoot: Boolean? = null
+
+    /**
+     * Gets the boot id, which is a value that changes every time the
+     * device is rebooted.
+     *
+     * For Android Oreo (SDK 24) and later, we can use the [BOOT_COUNT] global
+     * setting.
+     *
+     * For earlier versions of Android, we can use the Linux kernel's
+     * `/proc/sys/kernel/random/boot_id`.
+     *
+     * Adapted from https://github.com/instacart/truetime-android/pull/93
+     *
+     * @param context An application context
+     */
+    internal fun getBootId(context: Context): String {
+        if (Build.VERSION.SDK_INT > Build.VERSION_CODES.O) {
+            try {
+                return Settings.Global.getInt(
+                    context.getContentResolver(),
+                    Settings.Global.BOOT_COUNT
+                ).toString()
+            } catch (e: Settings.SettingNotFoundException) {
+                // pass through to old method
+            }
+        }
+
+        try {
+            val file = File("/proc/sys/kernel/random/boot_id")
+
+            val bootId = StringBuilder()
+
+            file.forEachLine {
+                bootId.append(it)
+            }
+
+            return bootId.toString()
+        } catch (e: IOException) {
+            // pass through to fallback method
+        }
```

Comment:
Is that the exception it throws on <= M too? Just wondering if it could be a `SecurityException`. That's what a bunch of things started throwing with M+ and runtime permissions.

Reply:
I couldn't actually reproduce the exception.  On `M` it passed, but I didn't try anything newer.

Reply:
Alright :)